### PR TITLE
[Merged by Bors] - feat: Construct a tripartite graph from its triangles

### PR DESCRIPTION
**Pull request by YaelDillies**
Both the lower bound of the Ruzsa-Szemerédi problem and the proof of the corners theorem define an auxiliary tripartite graph by specifying its triangles and showing that there are no more.

This PR provides the boilerplate for this construction. It greatly simplifies both proofs.

Co-authored-by: Bhavik Mehta <bhavik.mehta8@gmail.com>

---

- [x] depends on: #12526

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import Mathlib.Combinatorics.SimpleGraph.Triangle.Basic
+
+/-!
+# Construct a tripartite graph from its triangles
+
+This file contains the construction of a simple graph on `α ⊕ β ⊕ γ` from a list of triangles
+`(a, b, c)` (with `a` in the first summand, `b` in the second, `c` in the third).
```
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:12]: 'summand' is wrong here, you mean perhaps 'component'

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+We call
+* `t : Finset (α × β × γ)` the set of *triangle indices* (its elements are not triangles within the
+  graph but instead index them).
+* *explicit* a triangle of the constructed graph coming from a triangle index.
+* *accidental* a triangle of the constructed graph not coming from a triangle.
```
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:18]: 'not coming from a triangle index'

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+The two important properties of this construction are:
+* `SimpleGraph.TripartiteFromTriangles.ExplicitDisjoint`: Whether the explicit triangles are
+  edge-disjoint.
+* `SimpleGraph.TripartiteFromTriangles.NoAccidental`: Whether all triangles are explicit.
+
+This construction shows up unrelatingly twice in the theory of Roth numbers:
+* The lower bound of the Ruzsa-Szemerédi problem: From a Salem-Spencer set `s` on a commutative ring
```
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:26]: The terminology 'Salem-Spencer set' is not really used in the literature, and is perhaps confusing - better to say just 'a set without three-term arithmetic progressions'.

File: Mathlib/Data/Finset/Card.lean [hunk 4]
```diff
@@ -66,16 +66,14 @@ theorem card_le_card : s ⊆ t → s.card ≤ t.card :=
 theorem card_mono : Monotone (@card α) := by apply card_le_card
 #align finset.card_mono Finset.card_mono
 
-@[simp]
-theorem card_eq_zero : s.card = 0 ↔ s = ∅ :=
-  card_eq_zero.trans val_eq_zero
+@[simp] lemma card_eq_zero : s.card = 0 ↔ s = ∅ := card_eq_zero.trans val_eq_zero
```
Comment (TFBloom) [inline Mathlib/Data/Finset/Card.lean:69]: All the changes in this file have already been made in #12526 and should be removed from this PR.

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean [hunk 5]
```diff
@@ -28,23 +29,163 @@ This module defines and proves properties about triangles in simple graphs.
 
 ## TODO
 
-* Generalise `farFromTriangleFree` to other graphs, to state and prove the Graph Removal Lemma.
+* Generalise `FarFromTriangleFree` to other graphs, to state and prove the Graph Removal Lemma.
```
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean:32]: All the changes in this file have (I think) already been made in #12526 and should be removed from this PR.

**Review by TFBloom — CHANGES_REQUESTED** [review]

File: Mathlib/Data/Finset/Card.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:69]: (clarified in PM)

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean (hunk 5 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean:32]: (clarified in PM)

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean]: Opened #12682 to fix the existing mentions of the name.

**Review by TFBloom — APPROVED** [review]
This is now all good mathematics-wise, and I recommend it be merged (pending Lean approval).

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+/-- Predicate on the triangle indices for the explicit triangles to be edge-disjoint. -/
+class ExplicitDisjoint (t : Finset (α × β × γ)) : Prop where
+  inj₀ : ∀ ⦃a b c a'⦄, (a, b, c) ∈ t → (a', b, c) ∈ t → a = a'
+  inj₁ : ∀ ⦃a b c b'⦄, (a, b, c) ∈ t → (a, b', c) ∈ t → b = b'
+  inj₂ : ∀ ⦃a b c c'⦄, (a, b, c) ∈ t → (a, b, c') ∈ t → c = c'
+
+/-- Predicate on the triangle indices for there to be no accidental triangle.
+
+Note that we cheat a bit, since the exact translation of this informal description would have
+`(a', b', c') ∈ s` as a conclusion rather than `a = a' ∨ b = b' ∨ c = c'`. Those conditions are
```
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:122]: Should the s be t here?

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+@[simp] lemma card_triangles [NoAccidental t] : ((graph t).cliqueFinset 3).card = t.card :=
+by rw [cliqueFinset_eq_map, card_map]
+
+lemma farFromTriangleFree [ExplicitDisjoint t] {ε : 𝕜}
+    (ht : ε * ((Fintype.card α + Fintype.card β + Fintype.card γ) ^ 2 : ℕ) ≤ t.card) :
+  (graph t).FarFromTriangleFree ε :=
+farFromTriangleFree_of_disjoint_triangles (t.map toTriangle)
+  (map_subset_iff_subset_preimage.2 fun x hx ↦ by simpa using toTriangle_is3Clique hx)
+  (map_toTriangle_disjoint t) $  by simpa [add_assoc] using ht
```
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:227]:
```suggestion
  (map_toTriangle_disjoint t) $ by simpa [add_assoc] using ht
```

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+  refine ⟨?_, ?_, ?_⟩
+  · rintro ⟨rfl, rfl⟩
+    exact this rfl rfl (ExplicitDisjoint.inj₂ habc hxyz)
+  · rintro ⟨rfl, rfl⟩
+    exact this rfl (ExplicitDisjoint.inj₁ habc hxyz) rfl
+  · rintro ⟨rfl, rfl⟩
+    exact this (ExplicitDisjoint.inj₀ habc hxyz) rfl rfl
+
+lemma cliqueSet_eq_image [NoAccidental t] : (graph t).cliqueSet 3 = toTriangle '' t :=
+by ext; exact is3Clique_iff
```
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:208]: The prevailing style is to put the by at the end of the line (several times)

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]
Thanks, just some minor comments

**Review by jcommelin — COMMENTED** [review]
LGTM modulo Ruben's remarks

bors d+

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean (hunk 8 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:208]: As you might have noticed, ths file was hand-ported from Lean 3 😁

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**